src/Http/Http.Api/src/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,385 @@
+            //     return ExecuteValueTask(action(...), httpContext);
+            // }
+
+            // void return type
+
+            // ValueTask Invoke(HttpContext httpContext)
+            // {
+            //     action(...);
+            //     return default;
+            // }
+
+            var method = action.Method;
+
+            var needForm = false;
+            var needBody = false;
+            Type? bodyType = null;
+
+            // This argument represents the deserialized body returned from IHttpRequestReader
+            // when the method has a FromBody attribute declared
+
+            var args = new List<Expression>();
+
+            foreach (var parameter in method.GetParameters())
+            {
+                Expression paramterExpression = Expression.Default(parameter.ParameterType);
+
+                //if (parameter.FromQuery != null)
+                //{
+                //    var queryProperty = Expression.Property(httpRequestExpr, nameof(HttpRequest.Query));
+                //    paramterExpression = BindParamenter(queryProperty, parameter, parameter.FromQuery);
+                //}
+                //else if (parameter.FromHeader != null)
+                //{
+                //    var headersProperty = Expression.Property(httpRequestExpr, nameof(HttpRequest.Headers));
+                //    paramterExpression = BindParamenter(headersProperty, parameter, parameter.FromHeader);
+                //}
+                //else if (parameter.FromRoute != null)
+                //{
+                //    var routeValuesProperty = Expression.Property(httpRequestExpr, nameof(HttpRequest.RouteValues));
+                //    paramterExpression = BindParamenter(routeValuesProperty, parameter, parameter.FromRoute);
+                //}
+                //else if (parameter.FromCookie != null)
+                //{
+                //    var cookiesProperty = Expression.Property(httpRequestExpr, nameof(HttpRequest.Cookies));
+                //    paramterExpression = BindParamenter(cookiesProperty, parameter, parameter.FromCookie);
+                //}
+                //else if (parameter.FromServices)
+                //{
+                //    paramterExpression = Expression.Call(GetRequiredServiceMethodInfo.MakeGenericMethod(parameter.ParameterType), requestServicesExpr);
+                //}
+                //else if (parameter.FromForm != null)
+                //{
+                //    needForm = true;
+
+                //    var formProperty = Expression.Property(httpRequestExpr, nameof(HttpRequest.Form));
+                //    paramterExpression = BindParamenter(formProperty, parameter, parameter.FromForm);
+                //}
+                //else if (parameter.FromBody)
+                if (parameter.CustomAttributes.Any(a => typeof(IFromBodyAttribute).IsAssignableFrom(a.AttributeType)))
```

Comment:
We spent a bit of time trying to declutter ApiControllers by inferring default sources for parameters. The rules are fairly simple:

* Complex type -> FromBody
* Name appears as a route value -> FromRoute
* Everything else -> FromQuery

It helps declutter boiler plate code from your app. It would be crummy to lose that as part of this programming model.

Reply:
https://github.com/dotnet/aspnetcore/issues/30248